pkg/compact/compact.go

```diff
@@ -470,6 +474,87 @@ func (cg *Group) Resolution() int64 {
 	return cg.resolution
 }
 
+// should return the results/metrics of the planning simulation
+type PlanSim interface {
+	PlanProgressCalc(ctx context.Context) error
+}
+
+type DefaultPlanSim struct {
+	grouper               Grouper
+	planner               Planner
+	sy                    *Syncer
+	numberOfIterations    prometheus.Gauge
+	numberOfBlocksToMerge prometheus.Gauge
+}
+
+func NewDefaultPlanSim(grouper Grouper, planner Planner, sy *Syncer, reg prometheus.Registerer) *DefaultPlanSim {
+	return &DefaultPlanSim{
+		grouper: grouper,
+		planner: planner,
+		sy:      sy,
+		numberOfIterations: promauto.With(reg).NewGauge(prometheus.GaugeOpts{
+			Name: "thanos_number_of_compaction_iterations",
+			Help: "The number of compactions to do",
+		}),
+		numberOfBlocksToMerge: promauto.With(reg).NewGauge(prometheus.GaugeOpts{
+			Name: "thanos_number_of_blocks_to_merge",
+			Help: "The number of blocks to be merged",
+		}),
+	}
+}
+
+func (ps *DefaultPlanSim) PlanProgressCalc(ctx context.Context) error {
+	iterations := 0
+	blocksToMerge := 0
+
+	if err := ps.sy.SyncMetas(ctx); err != nil {
+		return errors.Wrapf(err, "could not sync metas")
+	}
+	originalMetas := ps.sy.Metas()
+
+	groups, err := ps.grouper.Groups(originalMetas)
+	if err != nil {
+		return errors.Wrapf(err, "could not group original metadata")
+	}
+
+	// figure out hasPlan and noPlan
+	for {
+		for _, g := range groups {
+			if len(g.IDs()) == 1 {
+				continue
+			}
+
+			// parameter should be of type tsdb.BlockMeta.meta
+			plan, err := ps.planner.Plan(ctx, g.Metadata())
+			if err != nil {
+				return errors.Wrapf(err, "could not plan")
+			}
+			if len(plan) == 0 {
+				continue
+			}
+			iterations++
+			ps.numberOfIterations.Set(float64(iterations))
```

Comment:
It is not necessary to update iterations here.
This should be updated after the whole simulation is done and the exposed metrics should reflect the actual iterations needed.